draft-ietf-protected-audience-key-value-services.md

```diff
@@ -0,0 +1,518 @@
+-   AEAD (Authenticated Encryption with Associated Data): AES-256-GCM, see
+    {{Section 7.3 of HPKE}}
+
+The server is repurposing the [OHTTP] encapsulation mechanism ({{Section 4.6 of OHTTP}}), new
+media types need to be defined:
+
+-   The OHTTP request media type is `message/ad-auction-trusted-signals-request`
+-   The OHTTP response media type is `message/ad-auction-trusted-signals-response`
+
+These media types are concatenated with other fields when creating the [HPKE] encryption context.
+
+### Message Framing {#framing}
+
+Before encryption and after decryption, the requests and responses have the following framing:
+
+| Byte     | 0         | 0             | 1 to 4   | 5 to Size+4       | Size+5 to end   |
+| -------- | --------- | ------------- | -------- | ----------------- | --------------- |
+| Bits     | 7-2       | 1-0           | \*       | \*                | \*              |
+| -------- | --------- | ------------- | -------- | ----------------- | --------------- |
+| Contents | Unused    | Compression   | Size     | Request Payload   | Padding         |
+
+The request/response is framed with this 5 byte header.
+
+The first byte is the format+compression byte. The lower 2 bits specify the format and compression ({{compression}}).
+The higher 6 bits are currently unused.
+
+The following 4 bytes are the length of the request message in network byte order.
+
+Then the request is zero padded to a set of pre-configured lengths.
+Padding is applied with sizes as multiples of 2^n KBs ranging from 0 to 2MB. So the valid response
+sizes will be `[0, 128B, 256B, 512B, 1KB, 2KB, 4KB, 8KB, 16KB, 32KB, 64KB, 128KB, 256KB, 512KB, 1MB,
+2MB]`.
+
+### Format+compression byte {#compression}
+
+| Compression | Description                    |
+| :---------: | :----------------------------- |
+|    0x00     | [CBOR], no compression         |
+|    0x01     | [CBOR], compressed in [Brotli] |
+|    0x02     | [CBOR], compressed in [GZIP]   |
+
+For requests, the byte value is 0x00. For responses, the byte value depends on the
+`acceptCompression` (see {{request-schema}}) field in the request and the server behavior.
+
+## Core Request Data {#request}
+
+Requests are not compressed and have a tree-like hierarchy:
+
+-   Each request contains one or more partitions. Each partition is a collection of keys that can be
+    processed together by the service without any potential privacy leakage. Keys from one interest
+    group must be in the same partition. Keys from different interest groups with the same joining
+    site (see https://wicg.github.io/turtledove/#joining-interest-groups) may or may not be in the same partition.
+-   Each partition contains one or more key groups. Each key group has its unique attributes among
+    all key groups in the partition. The attributes are represented by a list of `tags`. Besides
+    tags, the key group contains a list of keys to look up.
```

Comment:
I find this really hard to understand.  I don't understand how each key group is "unique".  What is unique about it? is this a requirement of the key group or just something the client does?
I don't understand what the "tags" are for.  I think we need some motivation.
I don't understand what it means "to look up" a key.  Can we explain what this means and what it is looked up in?

Reply:
Changed the wording.